Himani Parikh - 1322085

CSCI 636 - Big Data Analytics

Hands on 9.2 : Quiz

***Install pyspark***

In [ ]:
!pip install pyspark

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 316.9/316.9 MB 2.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for pyspark: filename=pyspark-3.5.0-py2.py3-none-any.whl size=317425344 sha256=3843d864aecdd051d21e523a68e42f5df1d3c9e44a88308723f48707c3fd4732
  Stored in directory: /root/.cache/pip/wheels/41/4e/10/c2cf2467f71c678cfc8a6b9ac9241e5e44a01940da8fbb17fc
Successfully built pyspark


In [ ]:
from pyspark import SparkContext
from pyspark.sql import SparkSession

spark = SparkSession\
        .builder\
        .master('local[*]')\
        .getOrCreate()
sc = spark.sparkContext

***Mount Google Drive***

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# change the directory to big-data-colab folder
%cd /content/drive/My Drive/Colab Notebooks/BigDataColabHimaniParikh

#pwd
!ls

/content/drive/My Drive/Colab Notebooks/BigDataColabHimaniParikh
2.DataCopyHimani.ipynb			      out_Datajoin
2_WordCountHimani.ipynb			      outDatajoin2
3.1_WordGrepHimani.ipynb		      out_grep
3.2_WordMedianHimani.ipynb		      out_testfile
4.MR-DataJoinHimani.ipynb		      out_testfile1
4.MR-WordCountHimani.ipynb		      out_testfile2
5.MR-DataJoin2Himani.ipynb		      out_wordmedian
6.MR-WordCountReducerHimani.ipynb	      SparkOutput
7.Spark-WordCountHimani.ipynb		      SqlDataSet
8.Spark-SQLHimani.ipynb			      testfile1
8.Spark-StreamingHimani.ipynb		      testfile2
9.Spark-Handling-missing-values-Himani.ipynb  WeatherDataset
BigDataHimani.ipynb			      wordcount_mapper.py
Data					      wordcount_num0_output.txt
Data_join				      wordcount_num2_output.txt
Data_join2				      wordcount_reducer.py
out


In [ ]:
# change the directory to big-data-colab folder
%cd /content/drive/My Drive/Colab Notebooks/BigDataColabHimaniParikh/WeatherDataset

!ls

/content/drive/My Drive/Colab Notebooks/BigDataColabHimaniParikh/WeatherDataset
daily_weather.csv


***Load classes and daily weather data.***

In [ ]:
from pyspark.sql import SQLContext
sqlsc = SQLContext(sc)

df = sqlsc.read.load('/content/drive/My Drive/Colab Notebooks/BigDataColabHimaniParikh/WeatherDataset/daily_weather.csv',
                     format='com.databricks.spark.csv',
                     header='true',
                     inferSchema='true')

/usr/local/lib/python3.10/dist-packages/pyspark/sql/context.py:113: FutureWarning: Deprecated in 3.0.0. Use SparkSession.builder.getOrCreate() instead.
  warnings.warn(


In [ ]:
df.describe().toPandas().transpose()

0                    1                   2  \
summary                 count                 mean              stddev   
number                   1095                547.0  316.24357700987383   
air_pressure_9am         1092    918.8825513138094   3.184161180386833   
air_temp_9am             1090    64.93300141287072  11.175514003175877   
avg_wind_direction_9am   1091    142.2355107005759   69.13785928889189   
avg_wind_speed_9am       1092     5.50828424225493  4.5528134655317185   
max_wind_direction_9am   1092   148.95351796516923   67.23801294602953   
max_wind_speed_9am       1091    7.019513529175272   5.598209170780958   
rain_accumulation_9am    1089  0.20307895225211126  1.5939521253574893   
rain_duration_9am        1092    294.1080522756142  1598.0787786601481   
relative_humidity_9am    1095    34.24140205923536  25.472066802250055   
relative_humidity_3pm    1095    35.34472714825898  22.524079453587273   

                                         3                   4  
summary                                min                 max  
number                                   0                1094  
air_pressure_9am         907.9900000000024   929.3200000000012  
air_temp_9am            36.752000000000685   98.90599999999992  
avg_wind_direction_9am  15.500000000000046               343.4  
avg_wind_speed_9am        0.69345139999974  23.554978199999763  
max_wind_direction_9am   28.89999999999991  312.19999999999993  
max_wind_speed_9am      1.1855782000000479   29.84077959999996  
rain_accumulation_9am                  0.0   24.01999999999907  
rain_duration_9am                      0.0             17704.0  
relative_humidity_9am    6.090000000001012    92.6200000000002  
relative_humidity_3pm   5.3000000000006855    92.2500000000003

In [ ]:
df.describe().show()

+-------+------------------+-----------------+------------------+----------------------+------------------+----------------------+------------------+---------------------+------------------+---------------------+---------------------+
|summary|            number| air_pressure_9am|      air_temp_9am|avg_wind_direction_9am|avg_wind_speed_9am|max_wind_direction_9am|max_wind_speed_9am|rain_accumulation_9am| rain_duration_9am|relative_humidity_9am|relative_humidity_3pm|
+-------+------------------+-----------------+------------------+----------------------+------------------+----------------------+------------------+---------------------+------------------+---------------------+---------------------+
|  count|              1095|             1092|              1090|                  1091|              1092|                  1092|              1091|                 1089|              1092|                 1095|                 1095|
|   mean|             547.0|918.8825513138094| 64.9330014128

In [ ]:
df.count()

1095

***Question 1:***  If we remove all missing values from the data, how many air pressure at 9am
measurements have values between 911.736 and 914.67?

A. 77

B. 287

C. 80

D. 96

***Answer: A***

In [ ]:
removeAllDF = df.na.drop()

In [ ]:
removeAllDF.count()

1064

In [ ]:
removeAllDF.describe(['air_temp_9am']).show()

+-------+------------------+
|summary|      air_temp_9am|
+-------+------------------+
|  count|              1064|
|   mean| 65.02260949558733|
| stddev|11.168033449415704|
|    min|36.752000000000685|
|    max| 98.90599999999992|
+-------+------------------+



In [ ]:
result = removeAllDF[(removeAllDF['air_pressure_9am'] > 911.736) &
            (removeAllDF['air_pressure_9am'] < 914.67)]

In [ ]:
result.count()

77

***Question 2***. If we impute the missing values with the minimum value, how many air
temperature at 9am measurements are less than 42.292?

A. 28

B. 23

C. 1

D. 5

***Answer: A - 28***

In [ ]:
from pyspark.sql.functions import min
imputeDf = df

In [ ]:
for x in imputeDf.columns:
  minValue = removeAllDF.agg(min(x)).first()[0]
  print(x, minValue)
  imputeDf =  imputeDf.na.fill(minValue, [x])

number 0
air_pressure_9am 907.9900000000024
air_temp_9am 36.752000000000685
avg_wind_direction_9am 15.500000000000046
avg_wind_speed_9am 0.69345139999974
max_wind_direction_9am 28.89999999999991
max_wind_speed_9am 1.1855782000000479
rain_accumulation_9am 0.0
rain_duration_9am 0.0
relative_humidity_9am 6.090000000001012
relative_humidity_3pm 5.3000000000006855


In [ ]:
result1 = imputeDf[(imputeDf['air_temp_9am'] < 42.292)]

In [ ]:
result1.count()

28

***Question 3.*** How many samples have missing values for air_pressure_9am?

A. 3

B. 5

C. 1092

D. 0

A***nswer: C-1092***

In [ ]:
df.describe(['air_pressure_9am']).show()
imputeDf.describe(['air_pressure_9am']).show()

+-------+-----------------+
|summary| air_pressure_9am|
+-------+-----------------+
|  count|             1092|
|   mean|918.8825513138094|
| stddev|3.184161180386833|
|    min|907.9900000000024|
|    max|929.3200000000012|
+-------+-----------------+

+-------+------------------+
|summary|  air_pressure_9am|
+-------+------------------+
|  count|              1095|
|   mean| 918.8527087074701|
| stddev|3.2304097463057584|
|    min| 907.9900000000024|
|    max| 929.3200000000012|
+-------+------------------+



***Question: 4***  Which column in the weather dataset has the most number of missing values?

A. rain_accumulation_9am

B. number

C. They are all the same

D. air_temp_9am

***Answer: B Number***

In [ ]:
from pyspark.sql.functions import max
missingDf = df

for x in missingDf.columns:
  maxValue = df.agg(max(x)).first()[0]
  print(x, maxValue)

number 1094
air_pressure_9am 929.3200000000012
air_temp_9am 98.90599999999992
avg_wind_direction_9am 343.4
avg_wind_speed_9am 23.554978199999763
max_wind_direction_9am 312.19999999999993
max_wind_speed_9am 29.84077959999996
rain_accumulation_9am 24.01999999999907
rain_duration_9am 17704.0
relative_humidity_9am 92.6200000000002
relative_humidity_3pm 92.2500000000003
